### PR TITLE
Remove usage of AddClassHandler that returns new lambda per invocation.

### PR DESCRIPTION
**Pull request by MarchingCube**
## What does the pull request do?

Cleanup codebase to always use `AddClassHandler` variant that does not force extra allocations per value change.

## What is the current behavior?

Every time value changes we allocate new delegate per subscriber. This quickly adds up for frequent events (input for example).

Sample trace from ControlCatalog:
image

## What is the updated/expected behavior with this PR?

Sample trace after this change:
image

## How was the solution implemented (if it's not obvious)?

Properties already had proper handler, for events I added a new one that does not force allocations. Marked allocating `AddClassHandler` methods obsolete.

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

## Fixed issues

Part of https://github.com/AvaloniaUI/Avalonia/issues/2919

### CODE REVIEW

File: src/Avalonia.Interactivity/RoutedEvent.cs [hunk 1]
```diff
@@ -113,24 +113,38 @@ public RoutedEvent(string name, RoutingStrategies routingStrategies, Type ownerT
             Contract.Requires<ArgumentNullException>(ownerType != null);
         }
 
+        [Obsolete("Use overload taking Action<TTarget, TEventArgs>.")]
```
Comment (jkoritzinsky) [inline src/Avalonia.Interactivity/RoutedEvent.cs]: I don't know if we want to mark this obsolete but I don't really feel super strongly about it.

**Review by jkoritzinsky — APPROVED** [review]

File: src/Avalonia.Interactivity/RoutedEvent.cs (hunk 1 above)
Comment (MarchingCube) [reply src/Avalonia.Interactivity/RoutedEvent.cs]: I originally thought about removing this completely as I see no value in this method, but I thought that somebody might be using it so I made it obsolete. Do you think that I should leave this in non-obsolete state?

File: src/Avalonia.Base/AvaloniaObjectExtensions.cs [hunk 2]
```diff
@@ -192,9 +192,9 @@ public static IObservable<T> GetObservable<T>(this IAvaloniaObject o, AvaloniaPr
         {
             return observable.Subscribe(e =>
             {
-                if (e.Sender is TTarget)
+                if (e.Sender is TTarget target)
```
Comment (Gillibald) [inline src/Avalonia.Base/AvaloniaObjectExtensions.cs]: Maybe a dumb question but when is the target not the sender? Isn't the observable always the sender? Do we need to pass the target to the action?

Comment (MarchingCube) [reply src/Avalonia.Base/AvaloniaObjectExtensions.cs]: Sender in this case is the control that is raising the event. And there is a mismatch sometimes so `e.Sender` might not be `TTarget` especially for properties that are defined somewhere and then we use `AddOwner` to use it somewhere else. `AddOwner` does not require objects to be derived from each other so in some cases there is a mismatch. 

For example this happens to `Content` property that is defined on `ContentControl`, but we have `ContentPresenter` using it and subscribing via `AddClassHandler` so it ends up getting a `Window` sometimes, which in this case will end up with not being a `ContentPresenter` so the handler is not invoked.

File: src/Avalonia.Controls.DataGrid/DataGrid.cs [hunk 3]
```diff
@@ -723,29 +723,29 @@ static DataGrid()
...
-            SelectionModeProperty.Changed.AddClassHandler<DataGrid>(x => x.OnSelectionModeChanged);
-            VerticalGridLinesBrushProperty.Changed.AddClassHandler<DataGrid>(x => x.OnVerticalGridLinesBrushChanged);
-            SelectedIndexProperty.Changed.AddClassHandler<DataGrid>(x => x.OnSelectedIndexChanged);
-            SelectedItemProperty.Changed.AddClassHandler<DataGrid>(x => x.OnSelectedItemChanged);
-            IsEnabledProperty.Changed.AddClassHandler<DataGrid>(x => x.DataGrid_IsEnabledChanged);
-            AreRowGroupHeadersFrozenProperty.Changed.AddClassHandler<DataGrid>(x => x.OnAreRowGroupHeadersFrozenChanged);
-            RowDetailsTemplateProperty.Changed.AddClassHandler<DataGrid>(x => x.OnRowDetailsTemplateChanged);
-            RowDetailsVisibilityModeProperty.Changed.AddClassHandler<DataGrid>(x => x.OnRowDetailsVisibilityModeChanged);
-            AutoGenerateColumnsProperty.Changed.AddClassHandler<DataGrid>(x => x.OnAutoGenerateColumnsChanged);
+            ItemsProperty.Changed.AddClassHandler<DataGrid>((x,e) => x.OnItemsPropertyChanged(e));
```
Comment (lindexi) [inline src/Avalonia.Controls.DataGrid/DataGrid.cs]: Format issus

File: src/Avalonia.Controls/Presenters/ItemsPresenter.cs [hunk 4]
```diff
@@ -35,7 +35,7 @@ static ItemsPresenter()
                 KeyboardNavigationMode.Once);
 
             VirtualizationModeProperty.Changed
-                .AddClassHandler<ItemsPresenter>(x => x.VirtualizationModeChanged);
+                .AddClassHandler<ItemsPresenter>((x,e) => x.VirtualizationModeChanged(e));
```
Comment (lindexi) [inline src/Avalonia.Controls/Presenters/ItemsPresenter.cs]: Format issue

File: src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs [hunk 5]
```diff
@@ -73,7 +73,7 @@ public class ScrollContentPresenter : ContentPresenter, IPresenter, IScrollable
         static ScrollContentPresenter()
         {
             ClipToBoundsProperty.OverrideDefaultValue(typeof(ScrollContentPresenter), true);
-            ChildProperty.Changed.AddClassHandler<ScrollContentPresenter>(x => x.ChildChanged);
+            ChildProperty.Changed.AddClassHandler<ScrollContentPresenter>((x,e) => x.ChildChanged(e));
```
Comment (lindexi) [inline src/Avalonia.Controls/Presenters/ScrollContentPresenter.cs]: Format issue

File: src/Avalonia.Controls/Primitives/SelectingItemsControl.cs [hunk 6]
```diff
@@ -119,7 +119,7 @@ public class SelectingItemsControl : ItemsControl
         /// </summary>
         static SelectingItemsControl()
         {
-            IsSelectedChangedEvent.AddClassHandler<SelectingItemsControl>(x => x.ContainerSelectionChanged);
+            IsSelectedChangedEvent.AddClassHandler<SelectingItemsControl>((x,e) => x.ContainerSelectionChanged(e));
```
Comment (lindexi) [inline src/Avalonia.Controls/Primitives/SelectingItemsControl.cs]: Format issue

File: src/Avalonia.Controls/ProgressBar.cs [hunk 7]
```diff
@@ -2,6 +2,7 @@
 // Licensed under the MIT license. See licence.md file in the project root for full license information.
 
 
+using System;
```
Comment (lindexi) [inline src/Avalonia.Controls/ProgressBar.cs]: Useless code

**Review by lindexi — COMMENTED** [review]
Some format issue

File: src/Avalonia.Controls.DataGrid/DataGrid.cs (hunk 3 above)
Comment (grokys) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: What is the format issue?

Comment (lindexi) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: @grokys We will add space after `,`

Comment (MarchingCube) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: I am away till Monday, so I can’t fix it right now.

Comment (grokys) [reply src/Avalonia.Controls.DataGrid/DataGrid.cs]: Fixed it myself @MarchingCube.

**Pull request merged**